libs/algorithms/include/hpx/parallel/container_algorithms/for_each.hpp

```diff
@@ -84,24 +241,345 @@ namespace hpx { namespace parallel { inline namespace v1 {
+    ///                     will be invoked for each of the elements as a
+    ///                     projection operation before the actual predicate
+    ///                     \a is invoked.
+    ///
+    /// \returns  The \a for_each algorithm returns a
+    ///           \a hpx::future<FwdIter> if the execution policy is of
+    ///           type
+    ///           \a sequenced_task_policy or
+    ///           \a parallel_task_policy and returns \a FwdIter
+    ///           otherwise.
+    ///           It returns \a last.
+    ///
+    template <typename ExPolicy, typename FwdIter, typename Size, typename F,
+        typename Proj = util::projection_identity>
+    typename util::detail::algorithm_result<ExPolicy, FwdIter>::type for_each_n(
+        ExPolicy&& policy, FwdIter first, Size count, F&& f,
+        Proj&& proj = Proj());
+}}    // namespace hpx::ranges
+#else
+#include <hpx/config.hpp>
+#include <hpx/concepts/concepts.hpp>
+#include <hpx/iterator_support/range.hpp>
+#include <hpx/iterator_support/traits/is_range.hpp>
+
+#include <hpx/algorithms/traits/projected_range.hpp>
+#include <hpx/parallel/algorithms/for_each.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+
+#include <type_traits>
+#include <utility>
+
+namespace hpx { namespace parallel { inline namespace v1 {
+    // clang-format off
     template <typename ExPolicy, typename Rng, typename F,
         typename Proj = util::projection_identity,
-        HPX_CONCEPT_REQUIRES_(execution::is_execution_policy<ExPolicy>::value&&
-                hpx::traits::is_range<Rng>::value&& traits::is_projected_range<
-                    Proj, Rng>::value&& traits::is_indirect_callable<ExPolicy,
-                    F, traits::projected_range<Proj, Rng>>::value)>
+        HPX_CONCEPT_REQUIRES_(
+            hpx::parallel::execution::is_execution_policy<ExPolicy>::value &&
+            hpx::traits::is_range<Rng>::value &&
+            hpx::parallel::traits::is_projected_range<Proj, Rng>::value &&
+            hpx::parallel::traits::is_indirect_callable<ExPolicy, F,
+                hpx::parallel::traits::projected_range<Proj, Rng>>::value)>
+    // clang-format on
     typename util::detail::algorithm_result<ExPolicy,
         typename hpx::traits::range_iterator<Rng>::type>::type
-    for_each(ExPolicy&& policy, Rng&& rng, F&& f, Proj&& proj = Proj())
+        HPX_DEPRECATED_V(1, 6,
+            "hpx::parallel::for_each is deprecated, use "
+            "hpx::ranges::for_each instead")
+            for_each(ExPolicy&& policy, Rng&& rng, F&& f, Proj&& proj = Proj())
     {
         return for_each(std::forward<ExPolicy>(policy), hpx::util::begin(rng),
             hpx::util::end(rng), std::forward<F>(f), std::forward<Proj>(proj));
     }
 }}}    // namespace hpx::parallel::v1
+
+namespace hpx { namespace ranges {
+    template <typename I, typename F>
+    using for_each_result = in_fun_result<I, F>;
```

Comment:
Shouldn't this be `util::in_fun_result`?

Reply:
Mmh, it could be either this or `parallel::util::in_fun_result` because of https://github.com/STEllAR-GROUP/hpx/blob/0bbfb3c5c22954b324dad2f43d1fb259998e9e11/libs/algorithms/include/hpx/parallel/util/result_types.hpp#L132. Do you have a preference?

Reply:
Ahh, no - you're right. This is fine as it is.